allensdk/brain_observatory/behavior/running_processing.py

```diff
@@ -14,40 +16,378 @@ def calc_deriv(x, time):
+        # positive wraps: subtract from the previous value and add vmax
+        unwrapped_diff[pos_wrap_ix] = (
+            (vsig[pos_wrap_ix] + vmax) - vsig_last[pos_wrap_ix])
+    # negative: subtract vmax and the previous value
+    if len(neg_wrap_ix):
+        unwrapped_diff[neg_wrap_ix] = (
+            vsig[neg_wrap_ix] - (vsig_last[neg_wrap_ix] + vmax))
+    # Other indices, just compute straight diff from previous value
+    wrap_ix = np.concatenate((pos_wrap_ix, neg_wrap_ix))
+    other_ix = np.array(list(set(range(len(vsig_last))).difference(wrap_ix)))
+    unwrapped_diff[other_ix] = vsig[other_ix] - vsig_last[other_ix]
+    # Correct for wrap artifacts based on allowed `max_diff` value
+    # (fill with nan)
+    # Suppress warnings when comparing with nan values to reduce noise
+    with np.errstate(invalid='ignore'):
+        unwrapped_diff = np.where(
+            np.abs(unwrapped_diff) <= max_diff, unwrapped_diff, np.nan)
+    # Get nan indices to propogate to the cumulative sum (otherwise
+    # treated as 0)
+    unwrapped_nans = np.array(np.isnan(unwrapped_diff)).nonzero()
+    summed_diff = np.nancumsum(unwrapped_diff) + vsig[0]    # Add the baseline
+    summed_diff[unwrapped_nans] = np.nan
+    return summed_diff
+
+
+def _zscore_threshold_1d(data: np.ndarray,
+                         threshold: float = 5.0) -> np.ndarray:
+    """
+    Replace values in 1d array `data` that exceed `threshold` number
+    of SDs from the mean with NaN.
+    Parameters
+    ---------
+
+    Returns
+    -------
+    np.ndarray (1d)
+        A copy of `data` with values exceeding `threshold` SDs from
+        the mean replaced with NaN.
+    """
+    corrected_data = data.copy().astype("float")
+    scores = zscore(data, nan_policy="omit")
+    # Suppress warnings when comparing to nan values to reduce noise
+    with np.errstate(invalid='ignore'):
+        exceed = np.array(scores > threshold).nonzero()[0]
+    corrected_data[exceed] = np.nan
+    return corrected_data
+
+
+def get_running_df(data, time: np.ndarray, lowpass: bool = True):
+    """
+    Given the data from the behavior 'pkl' file object and a 1d
+    array of timestamps, compute the running speed. Returns a
+    dataframe with the raw voltage data as well as the computed speed
+    at each timestamp. By default, the running speed is filtered with
+    a 10 Hz Butterworth lowpass filter to remove artifacts caused by
+    the rotary encoder.
+
+    Parameters
+    ----------
+    data
+        Deserialized 'behavior pkl' file data
+    time: np.ndarray (1d)
+        Timestamps for running data measurements
+    lowpass: bool (default=True)
+        Whether to apply a 10Hz low-pass filter to the running speed
+        data.
+
+    Returns
+    -------
+    pd.DataFrame
+        Dataframe with an index of timestamps and the following
+        columns:
+            "speed": computed running speed
+            "dx": angular change, computed during data collection
```

Comment:
probably not part of this PR, but, this is not a good name for an angular quantity.

Reply:
I agree, but this has been the name for a long time and it would be a confusing change. Probably would have ramifications on the NWB API, etc.

I think it's supposed to be dtheta.